### PR TITLE
UI Tag

### PR DESCRIPTION
**Pull request by nhaagen**
Tags are a variant of Button.
I am unsure about the aria-label - should not buttons in general have those?
If not, I'll add them to the Tag only.

### CODE REVIEW

File: src/UI/templates/default/Button/button.less
Comment (alex40724) [inline src/UI/templates/default/Button/button.less]: Without hovering there is no visiual difference between clickable and non-clickable tags, right? I think this is worth mentioning, but I also think it is ok. I think it is a good idea that tags are a type of button.

File: src/UI/Component/Button/Tag.php
Comment (alex40724) [inline src/UI/Component/Button/Tag.php]: First I saw the constants for the relevance style classes. I think it is good to have this set of style classes. But why do we need to set an arbitrary background and foreground color. Where would these values come from? If possible we should try to stick to a fixed set of CSS classes instead imo.

**Comment by klees** [comment]
Hi @Amstutz,
if you have time to review this one tomorrow, Nils might be able to finish it on friday for JF on monday.
Regards!

**Comment by alex40724** [comment]
Thanks for the PR! Will a separate Tag List / Tag Cloud container be implemented for tags?

File: src/UI/Component/Button/Tag.php
Comment (nhaagen) [reply src/UI/Component/Button/Tag.php]: We have cases where tags are system-induced, i.e. where trainings are tagged with categories.
Those categories are prone to changes, so certain users can adjust and extend categories. This includes coloring. 
Although I'm all in for classes, this would mean adding CSS for every new category, so I have this parallel mechanism where colors can be stored in the database instead. I think, both ways can easily live next to each other, with the CSS-classes as the default way of using tags.

**Comment by nhaagen** [comment]
Hi @alex40724, thanks for looking into this. A tag list/cloud is perfectly feasible, and I certainly have ideas about it. However, for the time being, I am not going there for no actual need. As of now the tags can be rendered into any dom-container which is sufficient for displaying lists at the object. In the future I'd like to see a cloud with different font-sizes, overlapping tags, etc... This would then belong to listings, right?

File: src/UI/templates/default/Button/button.less
Comment (nhaagen) [reply src/UI/templates/default/Button/button.less]: I updated the docs.

**Comment by alex40724** [comment]
In general I support to integrate this in the existing Tagging service.

But may I asked, what triggered this development? What will it be it's first use case? We are in the need of some kind of "tag input" element. Is it linked to this development? Or just customer patches for now?

File: src/UI/Component/Button/Tag.php
Comment (alex40724) [reply src/UI/Component/Button/Tag.php]: We have some similar coloring in the current ILIAS calendar (setup your own calendar, choose a color by entering any color code). Personally I think it was a bad idea and would like prefer to get rid of this. See this comment: https://github.com/ILIAS-eLearning/ILIAS/pull/550#discussion_r121371083

Having a fixed set of colors in a "marker palette" or something similar, would move the color code decision from the end user back to the designer (palette would be part of the skin/style) where it belongs imo. End users would only choose an entry from the palette.

Comment (alex40724) [reply src/UI/Component/Button/Tag.php]: I also asked @Amstutz /Yvonne to try setup a palette with a maximum of what they think are **distinguishable** colors (for our cases like tags/calendar appointments/letter avatars). Beyond this point it just doesn't make any sense to add more colors.

Comment (nhaagen) [reply src/UI/Component/Button/Tag.php]: Got your point. This seems to be a general decision on how to deal with colors. 
Will you decide on that during JF?

Comment (alex40724) [reply src/UI/Component/Button/Tag.php]: Yes, I hope so.

**Comment by klees** [comment]
Hi @alex40724, we are implementing this for the society, based on a devision by the Advisory Council. Regards!

**Comment by matthiaskunkel** [comment]
Jour Fixe, June 19, 2017: Pull request accepted. We prefer a default palette for colours that are controlled by less variables to allow designers to set a CI compliant set of colours that are distinguishable easily. Additionally, we would like to allow users to set colours for personal purposes - if the component allows this (like already known from the Calendar).

**Comment by Amstutz** [comment]
@nhaagen , @klees Is this ready for a final review of the implementation before we merge this? I o

**Comment by klees** [comment]
Hi @Amstutz,
I'll do a final review of this now. I also think we did not talk about the rules in the JF, although @matthiaskunkel s comment tells something different. So I think we should talk about this on the next regular JF.
Regards!

File: src/UI/Component/Button/Tag.php
Comment (klees) [inline src/UI/Component/Button/Tag.php]: Typehint on `Color` please.

File: src/UI/Component/Button/Factory.php
Comment (klees) [inline src/UI/Component/Button/Factory.php]: Let's talk about `entities` instead of objects, since `objects` already needs to carry a lot of meaning from ILIAS and OOP.

File: src/UI/Component/Button/Factory.php
Comment (klees) [inline src/UI/Component/Button/Factory.php]: Please remove the "how are they rendered" part and replace it by a description of the mere visuals (i.e. "text on a colored background").

File: src/UI/Component/Button/Factory.php
Comment (klees) [inline src/UI/Component/Button/Factory.php]: Make that a rule.

File: src/UI/Component/Button/Factory.php
Comment (klees) [inline src/UI/Component/Button/Factory.php:198]: I don't think that having an action should be the default for the tag. Please introduce a `withAction` instead.

File: src/UI/Component/Button/Tag.php
Comment (klees) [inline src/UI/Component/Button/Tag.php]: Please do not introduce the CSS-classes here. Instead make this relevances trigger the rendering of different blocks from the template and introduce the classes there.

Comment (klees) [reply src/UI/Component/Button/Tag.php]: Also low and verylow are twisted.

Comment (klees) [reply src/UI/Component/Button/Tag.php]: Also please name the constants like "LOW_RELEVANCE". Bytes are cheap nowadays =)

File: src/UI/Component/Button/Tag.php
Comment (klees) [inline src/UI/Component/Button/Tag.php]: Nope, one of the constants, I think.

File: src/UI/Component/Button/Tag.php
Comment (klees) [inline src/UI/Component/Button/Tag.php]: Remove this class thing regarding the relevance in this interface.

File: src/UI/Component/Button/Tag.php
Comment (klees) [inline src/UI/Component/Button/Tag.php]: Let's just use `isDark` from `Color` and make this black or white internally.

File: src/UI/Component/Button/Tag.php
Comment (klees) [inline src/UI/Component/Button/Tag.php]: These are classes for the tags, not for CSS. This is similar to the `name` from the Icon.

File: src/UI/Component/Button/Tag.php
Comment (klees) [inline src/UI/Component/Button/Tag.php]: Just one `getClasses` will be enought. The rest should be done in the rendering.

File: src/UI/Component/Button/Tag.php
Comment (klees) [inline src/UI/Component/Button/Tag.php]: Remove whitespace. Whitespacebytes are expensive nowadays.

File: src/UI/Implementation/Component/Button/Factory.php
Comment (klees) [inline src/UI/Implementation/Component/Button/Factory.php]: Reindent with tabs, please.

File: src/UI/templates/default/Button/button.less
Comment (klees) [inline src/UI/templates/default/Button/button.less]: @Amstutz: Is this `desaturate` something we want to have? I vaguely remember some discussion with @catkrahl about problems that arise due to this.

File: src/UI/templates/default/Button/button.less
Comment (klees) [inline src/UI/templates/default/Button/button.less]: This shouldn't be here.

File: src/UI/templates/default/Button/tpl.tag.html
Comment (klees) [inline src/UI/templates/default/Button/tpl.tag.html]: Can we render this as `<button>`?

**Comment by Amstutz** [comment]
@klees : Thanks for the Info. I do mine as soon as you are done, to so we prevent commenting on the same things twice. However I think we did bring this up in the JF, so there is no need to discuss the interface again unless there are objections voiced to any part of the interface. Such objections can always be discussed for any of the accepted entries.

We should handle JF resources with care. They are very limited and quite expensive considering how many people are participating.

**Review by klees — CHANGES_REQUESTED** [review]
Please change according to comments. Thx!

File: src/UI/Component/Button/Tag.php
Comment (nhaagen) [reply src/UI/Component/Button/Tag.php]: I think, this should be done when using the component. Why put a limit on it? 
How would you like the suggested mechanism to kick in when a background-, but no foreground-color is defined?

File: src/UI/templates/default/Button/tpl.tag.html
Comment (nhaagen) [reply src/UI/templates/default/Button/tpl.tag.html]: Basically yes, but this is for all buttons.

File: src/UI/Component/Button/Factory.php
Comment (nhaagen) [reply src/UI/Component/Button/Factory.php:198]: Since tags are buttons, they should have an action by default.

**Review by klees — APPROVED** [review]

**Comment by klees** [comment]
LGTM, thx. Please update the data.json to make this ready for merge.

**Comment by klees** [comment]
@Amstutz Can this be merged (minus conflicts in data.json)? Nils could the update data.json one last time...

File: src/UI/Implementation/Component/Button/Tag.php
Comment (Amstutz) [inline src/UI/Implementation/Component/Button/Tag.php:115]: I do not like this function (devs should not mess with CSS classes and such things). However it is part of the agreed upon interface, so we leave it as is.

File: src/UI/templates/default/Button/button.less
Comment (Amstutz) [reply src/UI/templates/default/Button/button.less]: I can not answer that. I leave that to @catkrahl to decide. However, desaturate(@btn-tag-bg, 80%) and contrast(desaturate(@btn-tag-bg, 80%)) can not be set here. They must be moved to seperate less variables. See: https://github.com/ILIAS-eLearning/ILIAS/tree/trunk/templates#colors

**Review by Amstutz — CHANGES_REQUESTED** [review]
Looks good to me in general. Two things:
- Resolve the merge conflicts and check if the tests still pass afterwards.
- Move some less logic to variables.less (see comments). 

Also I might slightly refactor the way we define the colors in less after we introduce the colors concept currently in discussion (@klees is informed). However this can be merged after the above points are resolved.

**Review by Amstutz — APPROVED** [review]
Thx.

File: src/UI/Implementation/Component/Button/Tag.php
Comment (klees) [reply src/UI/Implementation/Component/Button/Tag.php:115]: Hi @Amstutz,
this argument is very sophistic, but: A the concept of classes of tags exists, even if they were not transferred to CSS classes, at least in standard ILIAS. The tagging mechanism knows **their** tags and **my** tags, we generalised this to classes. The fact that they are directly translated to CSS-classes is a detail of our implementation.
Regards!

**Pull request merged**
